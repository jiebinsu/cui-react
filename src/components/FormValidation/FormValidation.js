import React from "react";
import PropTypes from "prop-types";
import shortid from "shortid";

const FormValidation = ({
  heading,
  message,
  errorLinks,
  hasErrors,
  children,
}) => {
  if (!hasErrors) return null;

  return (
    <div
      className="coop-c-message coop-c-message--error"
      aria-labelledby="error-message-heading"
      tabIndex="-1"
      role="alert"
    >
      <h2 id="error-message-heading" className="coop-c-message__heading">
        {heading}
      </h2>
      <p className="coop-c-message__message">{message}</p>
      <ul className="coop-c-message__list">
        {errorLinks.map(({ href, text }) => (
          <li key={shortid.generate()}>
            <a className="coop-c-message__link coop-t-link" href={href}>
              {text}
            </a>
          </li>
        ))}
      </ul>
      {children}
    </div>
  );
};

FormValidation.defaultProps = {
  heading: "Thereâ€™s a problem",
  message: "Check the form. You must:",
  errorLinks: [],
  hasErrors: false,
  children: null,
};

FormValidation.propTypes = {
  heading: PropTypes.string,
  message: PropTypes.string,
  errorLinks: PropTypes.arrayOf(
    PropTypes.shape({
      href: PropTypes.string,
      text: PropTypes.string,
    })
  ),
  hasErrors: PropTypes.bool,
  children: PropTypes.node,
};

export default FormValidation;
