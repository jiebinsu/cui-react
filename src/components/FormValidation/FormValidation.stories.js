import React from "react";
import FormValidation from "./FormValidation";

export default {
  title: "Base/FormValidation",
  component: FormValidation,
};

const errorLinks = [
  {
    href: "#client-name",
    text: "enter your full name",
  },
  {
    href: "#client-email-address",
    text: "enter an email address",
  },
];

export const Basic = (args) => <FormValidation {...args} />;
Basic.args = {
  heading: "Thereâ€™s a problem",
  message: "Check the form. You must:",
  errorLinks,
  hasErrors: true,
};

export const CustomBody = () => (
  <FormValidation hasErrors>
    <p>Add custom body</p>
  </FormValidation>
);
