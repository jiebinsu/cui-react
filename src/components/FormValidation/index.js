import FormValidation from "./FormValidation";

export default FormValidation;
