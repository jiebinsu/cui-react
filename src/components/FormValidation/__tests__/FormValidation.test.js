import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import FormValidation from "../index";

describe("FormValidation", () => {
  it("should render correctly", () => {
    const wrapper = mount(<FormValidation hasErrors />);
    expect(() => wrapper.unmount()).not.toThrow();

    const tree = renderer.create(<FormValidation hasErrors />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should not render when hasErrors is false", () => {
    const tree = renderer.create(<FormValidation />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render with links", () => {
    const errorLinks = [
      {
        href: "#client-name",
        text: "enter your full name",
      },
      {
        href: "#client-email-address",
        text: "enter an email address",
      },
    ];
    const tree = renderer
      .create(
        <FormValidation
          heading="title"
          message="message"
          errorLinks={errorLinks}
          hasErrors
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
