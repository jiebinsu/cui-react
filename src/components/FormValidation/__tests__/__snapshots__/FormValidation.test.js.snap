// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormValidation should not render when hasErrors is false 1`] = `null`;

exports[`FormValidation should render correctly 1`] = `
<div
  aria-labelledby="error-message-heading"
  className="coop-c-message coop-c-message--error"
  role="alert"
  tabIndex="-1"
>
  <h2
    className="coop-c-message__heading"
    id="error-message-heading"
  >
    Thereâ€™s a problem
  </h2>
  <p
    className="coop-c-message__message"
  >
    Check the form. You must:
  </p>
  <ul
    className="coop-c-message__list"
  />
</div>
`;

exports[`FormValidation should render with links 1`] = `
<div
  aria-labelledby="error-message-heading"
  className="coop-c-message coop-c-message--error"
  role="alert"
  tabIndex="-1"
>
  <h2
    className="coop-c-message__heading"
    id="error-message-heading"
  >
    title
  </h2>
  <p
    className="coop-c-message__message"
  >
    message
  </p>
  <ul
    className="coop-c-message__list"
  >
    <li>
      <a
        className="coop-c-message__link coop-t-link"
        href="#client-name"
      >
        enter your full name
      </a>
    </li>
    <li>
      <a
        className="coop-c-message__link coop-t-link"
        href="#client-email-address"
      >
        enter an email address
      </a>
    </li>
  </ul>
</div>
`;
